include/sys/scheduler.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_SCHEDULER_H
+#define ZEPHYR_INCLUDE_SYS_SCHEDULER_H
+
+#include <kernel.h>
+#include <sys_clock.h>
+#include <wait_q.h>
+
+/*
+ * APIs for working with the Zephyr kernel scheduler. Intended for use in
+ * management of IPC objects, either in the core kernel or other IPC
+ * implemented by OS compatibility layers, providing basic wait/wake operations
+ * with spinlocks used for synchronization.
+ *
+ * These APIs are public and will be treated as contract, even if the
+ * underlying scheduler implementation changes.
+ */
+
+/**
+ * Callback function prototype for k_sched_wake_one/wake_all
+ *
+ * If a thread is woken up, the callback function is called with the thread
+ * as an argument. The sched_spinlock is held the entire time, ensuring that
+ * thread state doesn't suddenly change (from some nested IRQ or activity on
+ * another CPU) when the callback is modifying the thread's state.
+ *
+ * It is only inside these callbacks that it is safe to inspect or modify
+ * the thread that was woken up.
```

Comment:
It would be good to document the intended use of `obj` and `context` here.

Also to document what can and cannot be done within the callback.  From what I can tell with #29610 if the callback invokes an operation that's a reschedule point spinlocks held by the caller might be broken.

Reply:
the callback holds `sched_spinlock`, so the true limitation is that no other API that holds `sched_spinlock` can be called. Simply stating this would be extremely unclear to users, I'll try to express in a better way.

Reply:
Thanks.  So in this case the spinlock wouldn't be broken because the callback would block forever waiting to take `sched_spinlock`.

Could you take a look at #29610 and correct my understanding if it isn't true in general that uniprocessor spinlocks (like `irq_lock`) broken if the current thread is preempted as a result of making a higher priority thread read and  rescheduling?  Is it true in SMP?

If either is true I'll work on a documentation update, because currently we warn about that only for `irq_lock`.